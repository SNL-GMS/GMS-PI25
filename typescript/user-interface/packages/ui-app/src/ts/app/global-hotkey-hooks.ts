import { useHotkeys } from '@blueprintjs/core';
import type { ConfigurationTypes } from '@gms/common-model';
import { buildHotkeyConfigArray, useGetProcessingAnalystConfigurationQuery } from '@gms/ui-state';
import { useMemo } from 'react';

/**
 * @returns the prevent default configuration from processing config, or undefined if the query
 * has not returned data
 */
export const useGlobalPreventDefaultsConfigurations = ():
  | ConfigurationTypes.PreventDefaultConfig
  | undefined => {
  return useGetProcessingAnalystConfigurationQuery().data?.preventBrowserDefaults;
};

/**
 * Enables the prevent global defaults browser config using blueprint hotkeys.
 */
export function useGlobalPreventDefault() {
  const preventBrowserDefaultsConfig = useGlobalPreventDefaultsConfigurations();
  // Build the config array
  const configs = useMemo(() => {
    return Object.values(preventBrowserDefaultsConfig || {})
      .map(value => {
        return buildHotkeyConfigArray(value, e => e.preventDefault(), undefined, false, true);
      })
      .flatMap(item => item);
  }, [preventBrowserDefaultsConfig]);

  useHotkeys(configs);
}
