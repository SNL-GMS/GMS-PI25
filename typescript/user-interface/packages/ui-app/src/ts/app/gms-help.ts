/* eslint-disable no-param-reassign */
import { IS_NODE_ENV_PRODUCTION } from '@gms/common-util';
import { uiStateHelp } from '@gms/ui-state';
import { ServiceWorkerMessages } from '@gms/ui-workers';

const codeStyle = 'font-family: monospace; font-size: 1.2em;';
const textStyle = `font-family: inherit; font-size: 1.2em;`;
const h1Style = `font-family: inherit; font-size: 2em;`;
const h2Style = `font-family: inherit; font-size: 1.5em;`;

function toggleWorkboxLogs() {
  navigator.serviceWorker.controller.postMessage(ServiceWorkerMessages.toggleWorkboxLogs);
}

export function attachGmsHelp() {
  if (!IS_NODE_ENV_PRODUCTION) {
    (window as any).gmsHelp = (window as any).gmsHelp || {};
    (function setGmsHelp(gmsHelp) {
      gmsHelp.please = () => {
        // Use console.log instead of logger because we don't want this logged back up to a server, ever.
        // eslint-disable-next-line no-console
        console.log(
          `
%cgmsHelp
%c----------


  The gmsHelp object is available in development mode (only).
  It is an object containing various help functions. 

  %cHow do I use it?%c
  - - - - - - - -
  Just call any of the functions defined within it:

  %c  gmsHelp.please() %c
  

  %cHelp functions%c
  - - - - - - - -

  please ()
    Print help info. "commands", "help", and "ls" are all aliased to this help function.

  toggleWorkboxLogs ()
    Toggle the service worker (workbox) logs. This takes effect immediately, and affects the state of the service worker. 
  ${uiStateHelp.please()}

  `,
          h1Style,
          textStyle,
          h2Style,
          textStyle,
          codeStyle,
          textStyle,
          h2Style,
          textStyle
        );
      };

      // assign ui-app functions
      gmsHelp.toggleWorkboxLogs = toggleWorkboxLogs;

      // aliases
      gmsHelp.help = gmsHelp.please;
      gmsHelp.ls = gmsHelp.please;
      gmsHelp.commands = gmsHelp.please;
      gmsHelp.toggleServiceWorkerLogs = gmsHelp.toggleWorkboxLogs;
      gmsHelp.toggleSwLogs = gmsHelp.toggleWorkboxLogs;

      // assign help functions from other packages
      Object.assign(gmsHelp, uiStateHelp.getHelpFunctions() ?? {});

      // log the help
      gmsHelp.please();
    })((window as any).gmsHelp);
  }
}
