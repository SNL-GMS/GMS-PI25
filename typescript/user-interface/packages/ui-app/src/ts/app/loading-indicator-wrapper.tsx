import { LoadingBar } from '@gms/ui-core-components';
import { selectRequestStats, useAppSelector } from '@gms/ui-state';
import React from 'react';

/**
 * Creates a loading indicator if there are pending requests
 */
export function LoadingIndicatorWrapper() {
  const { completed, initiated } = useAppSelector(selectRequestStats);
  return <LoadingBar isLoading={completed < initiated} />;
}
