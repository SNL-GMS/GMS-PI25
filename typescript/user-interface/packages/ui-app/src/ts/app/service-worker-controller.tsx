import { Logger } from '@gms/common-util';
import { nonIdealStateWithNoSpinner, nonIdealStateWithSpinner } from '@gms/ui-core-components';
import { analystActions, useAppDispatch } from '@gms/ui-state';
import type { RequestTrackerMessage } from '@gms/ui-workers';
import { ServiceWorkerMessages } from '@gms/ui-workers';
import { registerServiceWorker } from '@gms/ui-workers/lib/register-service-worker';
import throttle from 'lodash/throttle';
import React from 'react';

import { ServiceWorkerAlert } from './service-worker-alert';

const logger = Logger.create('GMS_SW_LOG', process.env.GMS_SW_LOG);

const LOADING_BAR_UPDATE_THROTTLE_MS = 250;

/**
 * This creates a number of variables used in the management of the service worker
 * 1. isAlertOpen, a boolean which controls the state of the alert.
 * 2. refreshCallback, a referentially stable function that
 * is provided to the registerServiceWorker function, which triggers the alert popup in case the service worker
 * registration detects that a new service worker needs to be installed
 * 3. skipWaitingFunctionRef, which is a ref that will contain the returned callback that should refresh the page.
 * In this case, we use a ref so that we don't capture an old value.
 * 4. toggleAlert, which is a referentially stable function that toggles whether the alert is open or not.
 */
const useAlert = () => {
  const skipWaitingFunctionRef = React.useRef<() => void>();
  const [isAlertOpen, setIsAlertOpen] = React.useState<boolean>(false);
  const toggleAlert = React.useCallback(() => setIsAlertOpen(isOpen => !isOpen), []);
  const refreshCallback = React.useCallback(
    (cb: () => void) => {
      skipWaitingFunctionRef.current = cb;
      toggleAlert();
    },
    [toggleAlert]
  );
  return { isAlertOpen, refreshCallback, skipWaitingFunctionRef, toggleAlert };
};

/**
 * Creates throttled update functions that store and eventually dispatch redux changes with the collected request
 * tracking messages.
 *
 * @returns referentially stable functions:
 * `addPendingRequest` adds a request to the queue, and schedules a throttled call to update the state. This is both a leading and trailing throttled call.
 * `addCompletedRequest` adds a request to the queue, and schedules a throttled call to update the state. This is just a trailing throttled call.
 */
const useRequestTracking = () => {
  const dispatch = useAppDispatch();
  const pendingRequestsRef = React.useRef<RequestTrackerMessage[]>([]);
  const completedRequestsRef = React.useRef<RequestTrackerMessage[]>([]);
  // dependencies should be kept exhaustive, however, eslint can't see through the throttle
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const dispatchPendingRequests = React.useCallback(
    throttle(
      () => {
        dispatch(analystActions.trackPendingRequests(pendingRequestsRef.current));
        pendingRequestsRef.current = [];
      },
      LOADING_BAR_UPDATE_THROTTLE_MS,
      { leading: true, trailing: true }
    ),
    [dispatch]
  );
  // dependencies should be kept exhaustive, however, eslint can't see through the throttle
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const dispatchCompletedRequests = React.useCallback(
    throttle(() => {
      dispatch(analystActions.trackCompletedRequests(completedRequestsRef.current));
      completedRequestsRef.current = [];
    }, LOADING_BAR_UPDATE_THROTTLE_MS),
    [dispatch]
  );
  const addPendingRequest = React.useCallback(
    (req: RequestTrackerMessage) => {
      pendingRequestsRef.current.push(req);
      dispatchPendingRequests();
    },
    [dispatchPendingRequests]
  );
  const addCompletedRequest = React.useCallback(
    (req: RequestTrackerMessage) => {
      completedRequestsRef.current.push(req);
      dispatchCompletedRequests();
    },
    [dispatchCompletedRequests]
  );
  return {
    addPendingRequest,
    addCompletedRequest
  };
};

/**
 * Calls to register the service worker, and listens for a reply, indicating that it has started correctly.
 *
 * @param refreshCallback the callback function that should be called if we need to ask the user if we can refresh
 * @param setIsServiceWorkerRegistered the setter to disable the service worker registering non ideal state
 * @param setError to set whether an error has occurred
 */
const useServiceWorker = (
  refreshCallback: (cb: () => void) => void,
  setIsServiceWorkerRegistered: React.Dispatch<React.SetStateAction<boolean>>,
  setError: React.Dispatch<React.SetStateAction<false | MessageEvent<string>>>
) => {
  const { addPendingRequest, addCompletedRequest } = useRequestTracking();
  React.useEffect(() => {
    async function startTheEngines() {
      navigator.serviceWorker.addEventListener('message', message => {
        logger.debug('[main] received message from service worker', message);
        if (message.data === ServiceWorkerMessages.listenersActiveMessage) {
          setIsServiceWorkerRegistered(true);
        } else if (message.data === ServiceWorkerMessages.clientConnectedMessage) {
          // no-op
        } else if (message.data?.message === ServiceWorkerMessages.requestInitiated) {
          addPendingRequest(message.data);
        } else if (message.data?.message === ServiceWorkerMessages.requestCompleted) {
          addCompletedRequest(message.data);
        } else {
          logger.error(
            'Unknown service worker message type received. See the following error for message details.'
          );
          setError(message);
          throw new Error(message.data);
        }
      });
      await registerServiceWorker(refreshCallback);
      logger.info('[main] registered service worker');
      await navigator.serviceWorker.ready;
      logger.info('[main] service worker ready');
      if (navigator.serviceWorker.controller) {
        logger.info('[main] postMessage', ServiceWorkerMessages.clientConnectedMessage);
        navigator.serviceWorker.controller.postMessage(
          ServiceWorkerMessages.clientConnectedMessage
        );
      }
    }
    if (process.env.GMS_SW !== 'false') {
      startTheEngines().catch(e => logger.error(e));
    }
  }, [
    addCompletedRequest,
    addPendingRequest,
    refreshCallback,
    setError,
    setIsServiceWorkerRegistered
  ]);
};

function MaybeServiceWorkerNonIdealState({
  children,
  isError,
  isServiceWorkerRegistered
}: React.PropsWithChildren<{
  isError: MessageEvent<string> | undefined;
  isServiceWorkerRegistered: boolean;
}>) {
  if (process.env.GMS_SW !== 'false' && isError) {
    return nonIdealStateWithNoSpinner('Error Registering Service Worker', isError.data);
  }
  if (process.env.GMS_SW !== 'false' && !isServiceWorkerRegistered) {
    return nonIdealStateWithSpinner('Registering Service Worker', 'Please wait');
  }
  return children;
}

type ServiceWorkerControllerProps = React.PropsWithChildren<unknown>;

/**
 * Registers the service worker, and if it is not registered, returns a non ideal state until
 * we get a message from the service worker. If it has a service worker that sends the proper
 * message, this renders its children.
 */
export function ServiceWorkerController({ children }: ServiceWorkerControllerProps) {
  const [isError, setError] = React.useState<MessageEvent<string> | undefined>(undefined);
  const [isServiceWorkerRegistered, setIsServiceWorkerRegistered] = React.useState(false);
  const { isAlertOpen, refreshCallback, skipWaitingFunctionRef, toggleAlert } = useAlert();

  useServiceWorker(refreshCallback, setIsServiceWorkerRegistered, setError);

  return (
    <>
      {!process.env.GMS_DISABLE_SERVICE_WORKER_ALERT && (
        <ServiceWorkerAlert
          isAlertOpen={isAlertOpen}
          skipWaitingFunctionRef={skipWaitingFunctionRef}
          toggleAlert={toggleAlert}
        />
      )}
      <MaybeServiceWorkerNonIdealState
        isError={isError}
        isServiceWorkerRegistered={isServiceWorkerRegistered}
      >
        {children}
      </MaybeServiceWorkerNonIdealState>
    </>
  );
}
