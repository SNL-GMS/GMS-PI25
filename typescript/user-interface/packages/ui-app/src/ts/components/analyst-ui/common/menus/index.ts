export { HideStationMenuItem } from './hide-station-menu-item';
