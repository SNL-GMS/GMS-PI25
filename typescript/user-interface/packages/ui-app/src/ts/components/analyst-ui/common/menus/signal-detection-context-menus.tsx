import type { ImperativeContextMenuOpenFunc } from '@gms/ui-core-components';
import { useImperativeContextMenuCallback } from '@gms/ui-core-components';
import React from 'react';

import type { SignalDetectionDetailsProps } from '../dialogs/signal-detection-details/types';
import { SignalDetectionContextMenu } from './signal-detection-context-menu';
import type { SignalDetectionContextMenuContentProps } from './signal-detection-context-menu/signal-detection-context-menu-content';
import { SignalDetectionDetailsContextMenu } from './signal-detection-details-context-menu';

export interface SignalDetectionContextMenusCallbacks {
  readonly signalDetectionContextMenuCb: ImperativeContextMenuOpenFunc<
    SignalDetectionContextMenuContentProps
  >;
  readonly signalDetectionDetailsCb: ImperativeContextMenuOpenFunc<SignalDetectionDetailsProps>;
}

export type SignalDetectionContextMenusGetOpenCallbackFunc = (
  callbacks: SignalDetectionContextMenusCallbacks
) => void;

export interface SignalDetectionContextMenusProps {
  getOpenCallback: SignalDetectionContextMenusGetOpenCallbackFunc;
}

/**
 * Handles the display of the Signal Detection Context Menus their callbacks.
 *
 * @params props @see {@link SignalDetectionContextMenusGetOpenCallbackFunc}
 */
export const SignalDetectionContextMenus = React.memo(function SignalDetectionContextMenus(
  props: SignalDetectionContextMenusProps
): JSX.Element {
  const { getOpenCallback } = props;

  const [
    signalDetectionContextMenuCb,
    setSignalDetectionContextMenuCb
  ] = useImperativeContextMenuCallback<SignalDetectionContextMenuContentProps>();

  const [signalDetectionDetailsCb, setSignalDetectionDetailsCb] = useImperativeContextMenuCallback<
    SignalDetectionDetailsProps
  >();

  React.useEffect(() => {
    if (signalDetectionDetailsCb) {
      getOpenCallback({
        signalDetectionContextMenuCb,
        signalDetectionDetailsCb
      });
    }
  }, [getOpenCallback, signalDetectionContextMenuCb, signalDetectionDetailsCb]);

  return (
    <>
      <SignalDetectionContextMenu
        getOpenCallback={setSignalDetectionContextMenuCb}
        signalDetectionDetailsCb={signalDetectionDetailsCb}
      />
      <SignalDetectionDetailsContextMenu getOpenCallback={setSignalDetectionDetailsCb} />
    </>
  );
});
