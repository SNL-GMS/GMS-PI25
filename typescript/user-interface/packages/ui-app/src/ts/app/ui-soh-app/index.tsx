import '~css/soh-app.scss';

import { isDarkMode, replaceFavIcon } from '@gms/ui-util';
import { enableAllPlugins } from 'immer';
import * as JQuery from 'jquery';
import React from 'react';
import ReactDom from 'react-dom';
import { createRoot } from 'react-dom/client';

import { KeyCloakService } from '~app/authentication/gms-keycloak';
import { GMS_DISABLE_KEYCLOAK_AUTH } from '~env';

import { checkEnvConfiguration } from '../check-env-configuration';
import { checkUserAgent } from '../check-user-agent';
import { configureElectron } from '../configure-electron';
import { App } from './app';

// required for golden-layout
(window as any).React = React;
(window as any).ReactDOM = ReactDom;
(window as any).createRoot = createRoot;
(window as any).$ = JQuery;

window.onload = () => {
  checkEnvConfiguration();
  checkUserAgent();
  enableAllPlugins();

  // if the user is in dark mode, we replace the favicon with a lighter icon so it is visible
  if (isDarkMode()) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports, global-require, import/no-unresolved
    const logo = require('../../../../resources/images/gms-logo-favicon-dark.png');
    replaceFavIcon(logo);
  }

  const root = createRoot(document.getElementById('app'));
  const renderApp = () => root.render(<App />);
  if (GMS_DISABLE_KEYCLOAK_AUTH) {
    renderApp();
  } else KeyCloakService.callLogin(renderApp);
};

configureElectron();
