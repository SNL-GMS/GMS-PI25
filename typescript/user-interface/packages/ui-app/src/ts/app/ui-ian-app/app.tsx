import { HotkeysProvider, Intent, NonIdealState, Spinner } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Displays } from '@gms/common-model';
import { IanDisplays } from '@gms/common-model/lib/displays/types';
import { IS_MODE_IAN } from '@gms/common-util';
import { getStore, useHistoryHotKeys, useUiTheme, withReduxProvider } from '@gms/ui-state';
import { useGlobalHotkeyListener } from '@gms/ui-util/lib/ui-util/hot-key-util';
import React from 'react';
import { Provider } from 'react-redux';
import { HashRouter, Route, Routes } from 'react-router-dom';

import { AnalystUiComponents } from '~analyst-ui/';
import { AnalystCommandRegistrar } from '~analyst-ui/commands/analyst-command-registrar';
import { FilterManager } from '~analyst-ui/components/filters/filter-manager';
import { InteractionConsumer } from '~analyst-ui/interactions/interaction-consumer';
import { InteractionProvider } from '~analyst-ui/interactions/interaction-provider';
import { authenticator } from '~app/authentication';
import { useGlobalPreventDefault } from '~app/global-hotkey-hooks';
import { EffectiveNowTimeInitializer } from '~app/initializers/effective-time-now-initializer';
import { UIThemeWrapper } from '~app/initializers/ui-theme-wrapper';
import { LoadingIndicatorWrapper } from '~app/loading-indicator-wrapper';
import { ServiceWorkerController } from '~app/service-worker-controller';
import { ThemedToastContainer } from '~app/themed-toast-container';
import { UniqueComponent } from '~common-ui/components/unique-component';
import { CommonCommandRegistrar } from '~components/common-ui/commands';
import {
  CommandPalette,
  useToggleCommandPaletteVisibilityHotkeys
} from '~components/common-ui/components/command-palette';
import { LoadingScreen } from '~components/loading-screen';
import { LoginScreen } from '~components/login-screen';
import { ProtectedRoute } from '~components/protected-route';
import { GoldenLayoutContext } from '~components/workspace/components/golden-layout/types';
import { ReduxWorkspaceContainer as Workspace } from '~components/workspace/workspace-container';
import { GMS_DISABLE_KEYCLOAK_AUTH } from '~env';

import { createPopoutComponent } from '../create-popout-component';
import { glContextData } from './golden-layout-config';

const GMS = 'GMS';
const store = getStore();

/**
 * Wraps the component route.
 * Provides the required context providers to the component.
 *
 * @param Component the component route
 * @param props the props passed down from the route to the component
 * @param suppressPopinIcon true to force suppress the golden-layout popin icon
 */
function wrap(Component: any, name: string | undefined, props: any, suppressPopinIcon = false) {
  function WrappedComponent(p) {
    const [uiTheme] = useUiTheme();
    useGlobalHotkeyListener();
    useGlobalPreventDefault();

    useHistoryHotKeys();
    useToggleCommandPaletteVisibilityHotkeys();

    const unique = name ? (
      <UniqueComponent name={name}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <Component {...p} />
      </UniqueComponent>
    ) : (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <Component {...p} />
    );

    return (
      <ServiceWorkerController>
        <HotkeysProvider>
          <>
            <UIThemeWrapper key={uiTheme.name}>
              <CommandPalette>
                <LoadingIndicatorWrapper />
                {unique}
                <InteractionProvider>
                  <InteractionConsumer />
                </InteractionProvider>
                <CommonCommandRegistrar />
                <AnalystCommandRegistrar />
              </CommandPalette>
              <ThemedToastContainer />
            </UIThemeWrapper>
            <EffectiveNowTimeInitializer />
            <FilterManager />
          </>
        </HotkeysProvider>
      </ServiceWorkerController>
    );
  }

  return createPopoutComponent(withReduxProvider(WrappedComponent), props, suppressPopinIcon);
}

const WrappedWorkflow: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Workflow, IanDisplays.WORKFLOW, props);
const WrappedEvents: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Events, IanDisplays.EVENTS, props);
const WrappedFilters: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Filters, IanDisplays.FILTERS, props);
const WrappedWaveform: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Waveform, IanDisplays.WAVEFORM, props);
const WrappedMap: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.IANMap, IanDisplays.MAP, props);
const WrappedSignalDetections: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.SignalDetections, IanDisplays.SIGNAL_DETECTIONS, props);
const WrappedStationProperties: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.StationProperties, IanDisplays.STATION_PROPERTIES, props);
const WrappedWorkspace: React.FC<unknown> = props => wrap(Workspace, undefined, props, true);
const WrappedAzimuthSlowness: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.AzimuthSlowness, IanDisplays.AZIMUTH_SLOWNESS, props);
const WrappedHistory: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.History, IanDisplays.HISTORY, props);

export function App(): React.ReactElement {
  const redirectRoute = GMS_DISABLE_KEYCLOAK_AUTH ? (
    <>
      <Route path="/login/:redirectUrl" element={<LoginScreen authenticator={authenticator} />} />
      <Route path="/login" element={<LoginScreen authenticator={authenticator} />} />
    </>
  ) : undefined;

  return !IS_MODE_IAN ? (
    <NonIdealState
      icon={IconNames.ERROR}
      action={<Spinner intent={Intent.DANGER} />}
      title="Invalid settings"
      description="Not configured for IAN mode - Please check settings"
    />
  ) : (
    <Provider store={store}>
      <HotkeysProvider>
        <HashRouter>
          {
            // ! CAUTION: when changing the route paths
            // The route paths must match the `golden-layout` component name for popout windows
            // For example, the component name `my-route` must have the route path of `my-route`
          }
          {
            // For performance use `render` which accepts a functional component
            // that won't get unnecessarily remounted like with component.
          }
          <Routes>
            {
              // Authentication
              // React router dropped support for optional params
              // so we need separate routes for login and login with a redirect
            }
            {redirectRoute}
            <Route
              path="/loading"
              element={
                <ProtectedRoute redirectPath="loading" title={`${GMS}: Loading...`}>
                  <LoadingScreen />
                </ProtectedRoute>
              }
            />
            {
              // Individual ProtectedRoutes
            }
            <Route
              path={`/${IanDisplays.WORKFLOW}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.WORKFLOW}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.WORKFLOW, `${GMS}: `)}
                >
                  <WrappedWorkflow />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.EVENTS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.EVENTS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.EVENTS, `${GMS}: `)}
                >
                  <WrappedEvents />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.FILTERS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.FILTERS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.FILTERS, `${GMS}: `)}
                >
                  <WrappedFilters />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.WAVEFORM}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.WAVEFORM}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.WAVEFORM, `${GMS}: `)}
                >
                  <WrappedWaveform />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.MAP}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.MAP}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.MAP, `${GMS}: `)}
                >
                  <WrappedMap />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.SIGNAL_DETECTIONS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.SIGNAL_DETECTIONS}
                  title={Displays.toDisplayTitle(
                    Displays.IanDisplays.SIGNAL_DETECTIONS,
                    `${GMS}: `
                  )}
                >
                  <WrappedSignalDetections />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.AZIMUTH_SLOWNESS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.AZIMUTH_SLOWNESS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.AZIMUTH_SLOWNESS, `${GMS}: `)}
                >
                  <WrappedAzimuthSlowness />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.STATION_PROPERTIES}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.STATION_PROPERTIES}
                  title={Displays.toDisplayTitle(
                    Displays.IanDisplays.STATION_PROPERTIES,
                    `${GMS}: `
                  )}
                >
                  <WrappedStationProperties />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.HISTORY}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.HISTORY}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.HISTORY, `${GMS}: `)}
                >
                  <WrappedHistory />
                </ProtectedRoute>
              }
            />
            {
              // Workspace
            }
            <Route
              path="*"
              element={
                <ProtectedRoute title={`${GMS}: App Workspace`}>
                  <GoldenLayoutContext.Provider value={glContextData()}>
                    <WrappedWorkspace />
                  </GoldenLayoutContext.Provider>
                </ProtectedRoute>
              }
            />
          </Routes>
        </HashRouter>
      </HotkeysProvider>
    </Provider>
  );
}
