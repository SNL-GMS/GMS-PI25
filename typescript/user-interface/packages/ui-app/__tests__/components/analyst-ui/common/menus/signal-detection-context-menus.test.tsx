import { signalDetectionsData } from '@gms/common-model/__tests__/__data__';
import { AnalystWorkspaceTypes, getStore } from '@gms/ui-state/lib/app';
import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import type { SignalDetectionContextMenuContentProps } from '~analyst-ui/common/menus/signal-detection-context-menu/signal-detection-context-menu-content';
import type { SignalDetectionContextMenusCallbacks } from '~analyst-ui/common/menus/signal-detection-context-menus';
import { SignalDetectionContextMenus } from '~analyst-ui/common/menus/signal-detection-context-menus';

jest.mock('@gms/ui-state', () => {
  const actualRedux = jest.requireActual('@gms/ui-state');
  const mockDispatchFunc = jest.fn();
  const mockDispatch = () => mockDispatchFunc;
  const mockUseAppDispatch = jest.fn(mockDispatch);
  return {
    ...actualRedux,
    useAppDispatch: mockUseAppDispatch,
    useUpdateSignalDetection: jest.fn()
  };
});
describe('Signal Detection Context Menus', () => {
  it('exists', () => {
    expect(SignalDetectionContextMenus).toBeDefined();
  });

  it('renders Signal Detection menus', async () => {
    let callbacks: SignalDetectionContextMenusCallbacks;
    const store = getStore();

    const Display = function Display() {
      return (
        <div>
          <Provider store={store}>
            <SignalDetectionContextMenus
              getOpenCallback={callback => {
                callbacks = callback;
              }}
            />
          </Provider>
        </div>
      );
    };

    const container = await waitFor(() => render(<Display />));
    expect(container.container).toMatchSnapshot();

    const mouseEvent = ({
      nativeEvent: new MouseEvent('contextmenu', {
        clientX: 100,
        clientY: 100
      }),
      preventDefault: jest.fn(),
      shiftKey: true,
      stopPropagation: jest.fn()
    } as unknown) as React.MouseEvent;

    const props: SignalDetectionContextMenuContentProps = {
      keyPrefix: 'context-menu-test',
      signalDetectionDetailsCb: jest.fn(),
      measurementMode: {
        mode: AnalystWorkspaceTypes.WaveformDisplayMode.MEASUREMENT,
        entries: {
          [signalDetectionsData[0].id]: true
        }
      },
      setMeasurementModeEntries: jest.fn()
    };

    await waitFor(() => {
      callbacks.signalDetectionContextMenuCb(mouseEvent, { ...props });
      callbacks.signalDetectionDetailsCb(mouseEvent, {
        signalDetection: signalDetectionsData[0]
      });
      container.rerender(<Display />);
    });
    await waitFor(() => {
      expect(container.baseElement).toMatchSnapshot();
    });
  });
});
