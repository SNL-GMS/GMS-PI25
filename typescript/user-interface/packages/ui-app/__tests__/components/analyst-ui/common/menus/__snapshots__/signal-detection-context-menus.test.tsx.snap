// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signal Detection Context Menus renders Signal Detection menus 1`] = `
<div>
  <div>
    <div
      class="bp4-context-menu2"
    >
      <div />
    </div>
    <div
      class="bp4-context-menu2"
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Signal Detection Context Menus renders Signal Detection menus 2`] = `
<body
  class=""
>
  <div>
    <div>
      <div
        class="bp4-context-menu2"
      >
        <div />
      </div>
      <div
        class="bp4-context-menu2"
      >
        <div />
      </div>
    </div>
  </div>
  <div
    class="bp4-portal"
  >
    <div
      class="bp4-context-menu2-virtual-target"
    />
  </div>
  <div
    class="bp4-portal"
  >
    <div
      class="bp4-context-menu2-virtual-target"
    />
  </div>
</body>
`;
